src/main/scala/rps/Game.scala

```diff
@@ -5,66 +5,92 @@ import scala.util.Random
+    case object CpuWins extends GameResult
   }
 
   def play(): Unit = {
-    println(s"""Choose your weapon:
-    ${Weapon.Rock} - ${matchWeapon(Weapon.Rock)}
-    ${Weapon.Paper} - ${matchWeapon(Weapon.Paper)}
-    ${Weapon.Scissors} - ${matchWeapon(Weapon.Scissors)}
-    """)
+    val menu = Move.moves
+      .map(m => s"${Move.encode(m)} - ${matchMove(Some(m))}")
+      .reduce((acc, curr) => s"${acc}\n${curr}")
+    println("Choose your weapon:")
+    println(menu)
+
     val rawUserInput = readLine()
-    val userWeapon = rawUserInput
-    val cpuWeapon = generateComputerMove()
+    val userMove = Move.decode(rawUserInput)
+    val cpuMove = generateComputerMove()
 
-    val result = checkWinner(userWeapon, cpuWeapon)
+    val result = checkWinner(userMove, cpuMove)
 
-    println(s"You chose: ${matchWeapon(userWeapon)} | your opponent chose: ${matchWeapon(cpuWeapon)}")
-    println(announceWinner(result))
+    println(
+      s"You chose: ${matchMove(userMove)} | your opponent chose: ${matchMove(cpuMove)}"
+    )
+    println(announceResult(result))
   }
 
-  def matchWeapon(weapon: String): String = {
-    weapon match {
-      case Weapon.Rock     => "🪨 Rock"
-      case Weapon.Paper    => "📄 Paper"
-      case Weapon.Scissors => "✂️ Scissors"
-      case _               => "💩 An invalid weapon"
+  def matchMove(move: Option[Move]): String = {
+    move.orNull match {
+      case Move.Rock     => "🪨 Rock"
+      case Move.Paper    => "📄 Paper"
+      case Move.Scissors => "✂️ Scissors"
+      case _             => "💩 An invalid weapon"
     }
   }
 
-  def announceWinner(result: String): String = {
+  def announceResult(result: GameResult): String = {
     result match {
-      case Winner.Draw => "It's a draw ✍️"
-      case Winner.User => "You win 🎉"
-      case _           => "You lose 🤷"
+      case GameResult.Draw     => "It's a draw ✍️"
+      case GameResult.UserWins => "You win 🎉"
+      case _                   => "You lose 🤷"
     }
   }
 
-  def checkWinner(userMove: String, computerMove: String): String = {
-    (userMove, computerMove) match {
-      case (x, y) if x == y              => Winner.Draw
-      case (Weapon.Rock, Weapon.Scissors)  => Winner.User
-      case (Weapon.Paper, Weapon.Rock)     => Winner.User
-      case (Weapon.Scissors, Weapon.Paper) => Winner.User
-      case _                               => Winner.Cpu
+  def checkWinner(
+      userMove: Option[Move],
+      computerMove: Option[Move]
+  ): GameResult = {
+    (userMove.orNull, computerMove.orNull) match {
+      case (x, y) if x == y            => GameResult.Draw // BUG? Both user and cpu use invalid weapons = draw
+      case (Move.Rock, Move.Scissors)  => GameResult.UserWins
+      case (Move.Paper, Move.Rock)     => GameResult.UserWins
+      case (Move.Scissors, Move.Paper) => GameResult.UserWins
+      case _                           => GameResult.CpuWins
       // default case also handles an invalid weapon (e.g. user inputs "4")
```

Comment:
A questo punto potrebbe avere senso gestire il caso di mossa non valida con un messaggio a parte, senza produrre un risultato in quel caso.

Reply:
Ok, fatto =)